### ISSUE TITLE
Does not seem to work for me

### ISSUE THREAD

**Issue by juddisjudd** [hectorgimenez/koolo#8]
I will be in game, start bot and then it seems to click in 3 spots and that's it.

**Few questions:**
- F keys supported? so like putting f1-f8?
- Guide on using a sorc?

**Comment by hectorgimenez** [comment]
Hey @juddisjudd let me take a look. I've been just building it for myself and playing around with it and I decided to publish it at some point, this is why documentation is so poor. I will try to find some time to test it better and write some doc

**Comment by hectorgimenez** [comment]
@juddisjudd you can also check the log.txt file generated within the same directory, maybe there are some clue about what's going on

**Comment by RobotGaston** [comment]
I have EXACTLY the same problem as juddisjudd.

If I run it from the exe file.
In the game it makes 2-3 clicks with the char and stops.

On the screen and in the log it keep saying that the selected script is running....but char is standing still and not moving.

If i try to run it from the main.go (go run cmd/koolo/main.go)
I get many errors:

\go\pkg\mod\github.com\robotn\gohook@v0.40.0\event.go:51:10: undefined: addEvent
..\go\pkg\mod\github.com\robotn\gohook@v0.40.0\event.go:62:7: undefined: Start

...(some MORE lines of errors here)..

\go\pkg\mod\github.com\robotn\gohook@v0.40.0\event.go:129:4: too many errors
(#)github.com/go-vgo/robotgo
..\go\pkg\mod\github.com\go-vgo\robotgo@v1.0.0-rc1\img.go:92:40: undefined: Bitmap
..\go\pkg\mod\github.com\go-vgo\robotgo@v1.0.0-rc1\img.go:107:38: undefined: Bitmap

...(some MORE lines of errors here)..
..\go\pkg\mod\github.com\go-vgo\robotgo@v1.0.0-rc1\robotgo_fn_v1.go:29:9: undefined: MoveSmooth
..\go\pkg\mod\github.com\go-vgo\robotgo@v1.0.0-rc1\robotgo_fn_v1.go:29:9: too many errors

please a hit, where to look to make it go?
appreciate it.

**Comment by hectorgimenez** [comment]
Is mouse pointer trying to move? I mean, when char stops, if you move the mouse is the application trying to move it back to any specific position, or basically doing nothing? As far as I understood, the console application is still open all the time, right? (is not crashing).
Could you please share the log? I know logging it's very poor at this time, hope it can be improved in the future.

Additionally, could you let me know what version of Windows are you using?

@RobotGaston About the errors you are seeing when trying to compile it, looks like MinGW toolchain is not properly set in the system PATH, this is required for the robotgo C bindings, for example in my case I have it in `E:\mingw64` so my PATH entry is `E:\mingw64\bin`

**Comment by RobotGaston** [comment]
thank you hector, 

the mouse pointer is not trying to move. It just moves briefly 2-3 times in a fast pace, after I start an exe file and then It let mouse go, after 3 fast clicks it's not trying to move with mouse at all it just frees mouse for a user and not trying to pick it up again. At the same time the log window showing "the selected routine (Andariel) is running..." and I can freely pick up and use the mouse to do what I like. In short It goes like this: I click on exe,  In Game the program picking up the mouse, giving with it 3 fast moves and clicks and let the mouse go. Char is at full stop. Mouse is free after this. Application is saying "running" and doing nothing, It's not trying to move a mouse or pick it up again. Yes. Console is open and saying that it successfully running while in reality it is not doing anything in a game or with the mouse at all. It is not crashing, I'm closing it the mouse, without any troubles. while it keeps running.

the only change that was made with the config file is the path with diablo 2 lod location corrected. (The are many lines if it's ok with you I could post it in this topic) I need to find out how can I hide it under the spoiler...or if you're ok with the wall of text I could put it here. 

The only changed line in the config is looking like this "D2LoDPath: 'C:\DiabloII1.13c'" The rest of the file is left untouched. 

This is the log file:

2023-05-18T15:33:18.240+0300	INFO	internal/supervisor.go:113	Diablo II: Resurrected window detected, offsetX: 8 offsetY: 31. Game Area Size X: 1518 Y: 919
2023-05-18T15:33:18.789+0300	INFO	internal/supervisor.go:131	Starting Game #1. Run list: Andariel, Mephisto, Pindleskin, Summoner
2023-05-18T15:33:18.789+0300	DEBUG	internal/bot.go:45	Fetching map data...
2023-05-18T15:33:22.120+0300	DEBUG	internal/bot.go:47	Fetch completed	{"ms": 3331}
2023-05-18T15:33:22.120+0300	INFO	internal/bot.go:52	Running: Andariel
2023-05-18T15:33:53.977+0300	INFO	koolo/main.go:49	Shutting down... 

this last line  - 2023-05-18T15:33:53.977+0300	INFO	koolo/main.go:49	Shutting down...  is comes when I manually closing the running application with the cross click with my mouse.

Thank you with info about MinGW, this is bad for me as downloaded MinGW, extracted it, set the path in windows environments and still the program can't see it :( I have no Idea where I gone wrong here.... :( If there's only a way to check if the MinGW is set up correctly...I go and double check again...

**Comment by hectorgimenez** [comment]
Looks like you have it correctly configured now, main branch is not building properly at this point so the error you are seeing is "good". I will fix it when I have some free time.

Regarding the character not moving... I tried the latest version with a clean Windows 10 installation and it worked for me, my default keybindings are a bit weird, please take a look at the config.yaml file to be sure all the bindings are matching, also ensure to enable/disable CTA if you don't have it, check the number of belt rows, the character you set in the configuration, inventory locked items, etc... Also one thing you can try to force the bot to do some previous actions before starting the run (to check if they work properly), is to put some items in your inventory (to not locked slots) to force it to stash them, or empty your potion belt to force it to go to the vendor to get more pots. At least you will be able to check if it's able to move properly.

Other thing you can do is to set the `renderMap: true` option in the config file, this should not be enabled during normal use because it greatly reduces the performance, but is useful for path debugging, it will generate a file named `cg.png` in the root directory, so you can check where the bot was trying to go before getting freeze.

**Comment by RobotGaston** [comment]
Ahh I see..

thank you for the post.

I've tried everything that you suggested, step by step (make 1 change run a program - not worked turn off - making a new change) and ...nothing helped :)

Checked all the keys mapping to correspond the ones that are set in config.

Filed up unlocked stash space with items, 

empted the belt,

turned CTA off in config,

after it

changed renderMap: true

in all these steps the program behaviour was the same,- 2 fast click (character quickly moved in game down and left) and application dropping mouse with infinitely "running..." on a log window...

It looks like to me the problem for me hides somewhere EARLIER in code before the program starting all the "town belt inventory algorithms :("
Code is not reaching "town-character routine parts". It stops somewhere before them...but where....what causes it...

Just noticed.., when i changed renderMap: true
the program is NOT generating file cg.png in the root directory (I checked everywhere root + all other folders). This file simply is not appearing.

thank you again for trying to help, if it works for you I guess the problem is somewhere on my side with the OS settings or who knows...I'm starting to think it's not possible to get it running for me :( It's just strange that we both,- me and "juddisjudd" have the same application behaviour...maybe he managed to solve it...lmk :)

**Comment by hectorgimenez** [comment]
Thanks for the feedback, I tagged a new version, added some extra logs along with many additional changes, let me know if it works (or if it shows something extra in logs).

**Comment by RobotGaston** [comment]
thank you for a new branch. Well this is a massive improvement.

Program now is not dropping mouse and trying to do necessary things in a game. 
we r past the first described problem. it's resolved.
but I have MANY others now. )
still not working for me. details are progress. a LOT to write.... )

**Comment by hectorgimenez** [comment]
Ensure that your game resolution is set to 1280x720 (windowed mode), what resolution are you using in your desktop? I have a 4k monitor with hidpi mode, so for example my game resolution must be 1920x1080 to compensate the scale ratio.

`INFO internal/supervisor.go:114 Diablo II: Resurrected window detected, offsetX: -1 offsetY: 27. Game Area Size X: 1536 Y: 864`

This log line suggests that you are using a different game resolution.

**Comment by RobotGaston** [comment]
thank you, hectorgimenez

for pointing out to the window size, you where absolutely right,

wrong size of the window caused major of described problems, 
now, when the window is in 1280x720 (windowed mode) (it is strange btw. While the actual window size is 1280x720 - it is a resolution detected by the program, at the same time in game resolution in settings is set to 1280x900, when I set it to 1280x720 in game the actual window size is different (it is not 1280x720), don't bother with it's just I'll have to fix the actual window size myself to set it to correct size..) 

the program can finally run the task for me. It can detect Akara shop bottles - it can place cursor CORRECTLY ! on the items awesome. It can successfully select waypoint move to the target kill check loot and make a new game. 

I have a trouble with a game creation time :) my rig can't keep up with the program I need to increase game creating time by  3-4 seconds. The programs clicks on play-difficulty too fast for my pc - it does that while the screen still black - menu is loading and the game creation loop is not working for me - but this is problem with my gear. It's not that fast...to load game faster...

this is below is a raw data: (I'll keep it testing for updates) and might be ignored.

ok... we completed a run. We loaded into town. and now it's a stash management time.  It goes to stash opens it, browses alongs the Inlays and not stashing anything that I have in my inventory,

the items just keep siting  with my character. Inventory is not emptying for me. Some items like key for example it's manage to stash but big unidentified shield just left untouched. It is not thrown away, not id'eed not sold to Akara just siting occupying inventory (I checked the space is set to '1' in config - available to be cleared)

I need more testing time to look into how it interacts with stashing idiing vendoring.. it might be I have again somewhere something messed up. 

thank you ,hectorgimenez

**Comment by hectorgimenez** [comment]
Cool, glad to see that you managed to get it work. Feel free to open a new issue if you find any other problem, or a PR will be very welcome :)

**Comment by RobotGaston** [comment]
cheers, I'm enjoying watching it go :)

what I have noticed - 

static is not working on my sorc (it is configured and key is buinded accordingly in config, skill is available) sorc just ignoring  it. Not casting it. 

I checked "blizzard_sorceress" and this line is commented out, might it be a reason for it not working?

..........//if useStaticField {................

"The game creation" would be great to give it a little bit more time to create a game from menu +3 more seconds before execute routine play-difficulty. Mine game is not fast enough to load the menu sometimes (play-difficulty is executing while my screen still black and menu is not loaded) and It's missing the game creation this way :) But this is my problem and my old rig's...

also I'm very interested in companion function, how does it work?, If I have another character can It join the main Leader's game and give leader BO for example on waypoint then switch the window back to leader's and leader can go with the selected routine :) When the game is over 2'd acc leaving leaders game and waiting for a new game to be created to join and repeat routine. 

Also is it possible to use another sorceress build like a nova or fireball for example? with the current application by just changing skills in excising config file? Or it requires to make a new sorceress attack file for another skills? Will fireball skill for example work instead of blizzard with the current " atack " file?

and last )... after the fight is over it feels that there are items present that it needs to pick up :) and it's starting to draw a small tiny circles trying to "catch" an items and it can click on it only when the cursor is finally finds the object that it's trying to click. I mean "little circle" mode is the only available one, it just taking a lot of time to "draw a detecting" circles :) If it is than great. :) Just is there any way to speed the detecting process and placing cursor+click on an items faster? ) If no the no. :) 

this is just amazing.

**Comment by hectorgimenez** [comment]
Yeah, static field is not working properly, I want to refactor that part of the code and I will take a look at the static field thingy, but still didn't had time for it.

About the game creation delay... It would be nice if we can detect when the character selection screen is fully loaded, so we would not need to deal with static delays, but I'm not sure how to do it. Also not a big deal, if the game can not be created on the first attempt, will try again after few seconds.

And for the companion mode... it's very very stupid at this point. Basically requires from a Discord bot key to sync between different bot instances, the leader is creating the game and sharing via discord the game info and secondaries are joining the same game, they must be in the same act and they will wait until a portal from the leader is opened, then they will join and try to follow the leader until it leaves the game (and they will do the same). And start again. So it requires two computers or more (or VMs). Was more about doing baalruns or something like that for leveling.

About nova/fireball sorc, technically it will work if you just bind other spells to the blizzard keybinding. The thing with current blizzard_sorceress script is that is optimized for that, the blizzard. It might also work with nova, but all the attack strategy and rotation is based on blizzard one. So probably it would be better to create a new one to fine tune and adjust it to the nova rotation, distances, etc... it should be relatively easy to do if you want, you can start copyng the blizzard one (I don't have nova sorc)

About the item pickup... yeah it fails sometimes, but it's happening always? (it happens when the show item names option is enabled, because the hitbox for the items is offsetted, so it's better to disable it) the mouse trying to draw a circle it's basically to avoid when there is something blocking the bot to detect the item under the cursor, and because probably the game <> screen coords conversion is far for being perfect, so there is a small error ratio there, only noticiable when destination is very small, like item hitbox.

**Comment by RobotGaston** [comment]
thank you very much for your detail response read everything carefully with a great appreciation, hectorgimenez

I just can't enjoy enough to look at it working :) Great job. 

So.. it will try to make a game again, if the first game creation routine was lost in the loading screen....well this is GREAT, and completely fine. I haven't been patient enough to wait for a 2nd attempt to occur I guess :)

As continuing testing I noticed that the blizzard sorc sometimes... in RARE occasions (In an ancient tunnels for example or in lower curast) can get into a situation when there is an obstacle between her and the mob she's trying to attack (a pillar or a column) and she's getting herself stuck in firing spells in this obstacle...thinking that she's hitting the mob...(I don't know how or will she at all realise that,- the target is not getting hit and try to reposition itself for a new angle for attack...I personally solving it manually moving character into a new spot). And on countess also this might happen like in 20% rate that it can locate behind the wall, and the character in game shooting into the wall :) I don't mind this at all! (more than happy with everything) just an observation data if you'd like to have it. 

Thank you again for the advices, (and all the explanations about the mechanics, static, skills and leader-follower system) Greatly appreciate it and your time.

Now I can see more clearly what to do If I need game behaviour go the way I like them to go...which things are possible and which are not...

I'll maybe try to implement them myself and see how it goes, blizzard sorc is doing fine. 

I'm more than happy with the current state of how it works.

No problems here . Everything is resolved for me.

**Comment by RobotGaston** [comment]
More observation. on hammerdin now. :) Tested him for a bit...

there's a **bug** happening when you try to run games with him which is making it impossible for him to run in mode "set and forget" for me...

starting from the top:

in a config file for hammerding there's a dilemma... 

is section 
 -- Skills
  swapWeapon: w
we have this

and in section
  hammerdin:
    concentration: w
this..

I had to make a decision what to do and remapped concentration to another button in config leaving weapon swap on "w".
It's not important (hammerdin can do swap and use concentration correctly) just mention in case action and according button pressings of the "w" button are "hardcoded" somewhere. 

for me hammerdin.

is not using vigor while running in town.
hammerdin is not using redemption in fields.
and is not using cleansing either. 

Skills are configured in config and mapped with corresponded keys in game. 

  hammerdin:
    concentration: u
    holyShield: a
    vigor: g
    redemption: r
    cleansing: f

It doesn't matter much that vigor or redemption are not used.

the important thing is that Overall hammerdin is doing okay. He can run game after game after game. But at some point... and in random place he fails. 

He can sometimes have difficulty locating mephisto. It goes like this. Palading reached meph successfully. Pala moves cursor to an area where mephisto is, mephisto does a quick move far away from his location, and pala "looses him" with the mouse. He just stuck with mouse moves in an area where he's trying to locate mephisto, but the meph far long gone to another area where he fights with the merc alone...when pala is busy with the mouse. (not attacking the meph). Sorc does a tele to a meph at the middle of a fight. Maybe would be great to force hammer make a similar move. 

in ancient tunnels pala can experience same behaviour as a sorc - get stuck into column and cast hammers directly into it for a long time.

at travi consils, and some other places (flying creatures in arcane) overall positioning..he sometimes maybe 30% of a time is getting stuck in an attacking position missing with attacks  where hammers can't reach the mobs. Pala is placing himself into a position right on top of the mob when hammers can't hit them. He keeps spinning and spinning but all hammers are missing the target.  I believe a small forced check+change of a position would solve this. A small reposition. Tele stomp slightly back or to the left almost in place. In travi consuls could spawn in a very bad positions for hammes. Very close to the wall and ot forces pala to go there near them for atack to get stuck to the wall and makes him casting hammers right into the wall missing all his atacks...Changing pala's behaviour for consuls making him more aware of a surroundings would be just amazing...

Making pala to pick up consuls at a meph location would be awesome too.

and to the bug. 

Have no idea what causes it. Pala running fine... game after game... location after location then, suddenly, in the same location he successfully ran in a previous game he fails.
I mostly experienced it in a lower curast and at mephisto. 
Pala is in a location and doing what he supposed to do, then he stops on the screen and mouse cursor keeping moving within an area of a game screen and moves here and there, just cursor no clicks. It can go on for a minute then the cursor **leaving** the game window and app sends keypress out of the game window causing it to minimize to tray. If at this moment I hit pause and manually get back into the game window I find myself in a position where I can't click with the mouse. left or right clicks are not working. The keyboard works fine at the same moment.
This is solving with by restarting the app...

if only the according tweaks with hammerdin could be implemented that would be a saviour...

he's much tougher then a sorc, it's bugger to see him misses his strong hammers here and there and staying afk in meph fight letting merc to solve the problem :)
